cranelift/codegen/src/isa/x64/lower.rs

```diff
@@ -1343,27 +1343,71 @@ fn lower_insn_to_regs<C: LowerCtx<I = Inst>>(
             // - to compute the negated value, set all bits to 0 but the MSB to 1, and bit-XOR the
             // src with it.
             let output_ty = ty.unwrap();
-            let (val, opcode) = match output_ty {
-                F32 => match op {
-                    Opcode::Fabs => (0x7fffffff, SseOpcode::Andps),
-                    Opcode::Fneg => (0x80000000, SseOpcode::Xorps),
-                    _ => unreachable!(),
-                },
-                F64 => match op {
-                    Opcode::Fabs => (0x7fffffffffffffff, SseOpcode::Andpd),
-                    Opcode::Fneg => (0x8000000000000000, SseOpcode::Xorpd),
-                    _ => unreachable!(),
-                },
-                _ => panic!("unexpected type {:?} for Fabs", output_ty),
-            };
+            if !output_ty.is_vector() {
+                let (val, opcode) = match output_ty {
+                    F32 => match op {
+                        Opcode::Fabs => (0x7fffffff, SseOpcode::Andps),
+                        Opcode::Fneg => (0x80000000, SseOpcode::Xorps),
+                        _ => unreachable!(),
+                    },
+                    F64 => match op {
+                        Opcode::Fabs => (0x7fffffffffffffff, SseOpcode::Andpd),
+                        Opcode::Fneg => (0x8000000000000000, SseOpcode::Xorpd),
+                        _ => unreachable!(),
+                    },
+                    _ => panic!("unexpected type {:?} for Fabs", output_ty),
+                };
 
-            for inst in Inst::gen_constant(dst, val, output_ty, |reg_class, ty| {
-                ctx.alloc_tmp(reg_class, ty)
-            }) {
-                ctx.emit(inst);
-            }
+                for inst in Inst::gen_constant(dst, val, output_ty, |reg_class, ty| {
+                    ctx.alloc_tmp(reg_class, ty)
+                }) {
+                    ctx.emit(inst);
+                }
+
+                ctx.emit(Inst::xmm_rm_r(opcode, src, dst));
+            } else {
+                // Eventually vector constants should be available in `gen_constant` and this block
+                // can be merged with the one above (TODO).
+                if output_ty.bits() == 128 {
+                    // Move the `lhs` to the same register as `dst`; this may not emit an actual move
+                    // but ensures that the registers are the same to match x86's read-write operand
+                    // encoding.
```

Comment:
Hmm, we shouldn't probably duplicate this comment everywhere, since it's the regular way to do things in x86; we may add a file doc-comment at some point explaining this, maybe. It's fine to keep this comment until then.